gh-1006: `Menu.SetItemExtendedData`

MKadaner opened the pull request:
## Summary
Added Lua API `B=Menu.SetItemExtendedData([hDlg,][N,]T)`.

Also, updated `Edit.SyncEditorAndFindAllMenu.lua` to demonstrate usage of this API.

## References
- gh-1006

<!-- Please review the items on the PR checklist before submitting -->
## Checklist
- [X] I have followed the [contributing guidelines](https://github.com/FarGroup/FarManager/blob/master/CONTRIBUTING.md).
- [X] I have discussed this with project maintainers: gh-1006<br/>
If not checked, I accept that this work might be rejected in favor of a different great big ineffable plan.<br/>

<!-- Enter a more detailed description of the PR and any additional comments here -->
## Details
The implementation is ready for review, complete with the support in the Editor. Special thanks to @shmuz for the great help with passing Lua tables as parameters to native functions, and to @zg0 for the suggested improvements.

Now, one can write in Lua:

```Lua
Menu.SetItemExtendedData(FarDialogEvent.hDlg, SelectPos, { Line = 496, Position = 28, Length = 6 })
```

@rohitab, could you please test the new function with your macro?

@alabuzhev, could you please look at Editor and VMenu?

MKadaner committed:
gh-1006: Menu.SetItemExtendedData
1. gh-1006: Macro API to set menu item extended data: `Menu.SetItemExtendedData`.
2. Updated `Edit.SyncEditorAndFindAllMenu.lua` to demonstrate usage of this API.

rohitab commented:
> Now, one can write in Lua:
> 
> ```lua
> Menu.SetItemExtendedData(FarDialogEvent.hDlg, SelectPos, { Line = 496, Position = 28, Length = 6 })
> ```
> 
> @rohitab, could you please test the new function with your macro?

I couldn't get this to work. Maybe I'm not doing it right. After calling `Menu.SetItemExtendedData`, I don't see any change in the menu. The information displayed is exactly the same as before, i.e., the one displayed before the call. However, if I call `Menu.GetItemExtendedData` after setting the data, it returns the data that was set. If I now press `F4`, the selection is correct as well, and matches the `Line`, `Position` and `Length` that was passed to `Menu.SetItemExtendedData`.

alabuzhev commented on far/macroapi.cpp:
```diff
@@ -1204,31 +1251,62 @@ void KeyMacro::CallFar(intptr_t CheckCode, FarMacroCall* Data)
+				api.PushTable();
+				for (const auto& [Key, Value] : ExtendedData)
+				{
+					api.SetField(Key, Value, -3);
```
What is -3?
Maybe introduce some named constant or add a comment?

shmuz replied:
> What is -3?
Maybe introduce some named constant or add a comment?

It is table position in Lua stack.
Stack position is often expressed as a negative value relative to stack top.
When we want to set a key-value pair in a table we first push the key then push the value.
If initially the table was on Lua stack top its stack position was -1.
After key and value were pushed the table position became -3.

MKadaner replied:
Thank you for the explanation, @shmuz.

@alabuzhev, as you can see, it is not a magic constant. I'll add a comment.

alabuzhev commented on plugins/luamacro/luafar/lf_luamacro.c:
```diff
@@ -201,6 +201,22 @@ static void WINAPI MacroCallFarCallback(void *Data, struct FarMacroValue *Val, s
 			break;
 
+		case FMVT_NEXT:
+			if (pos >= 1 && pos <= top - 1 && lua_istable(L, pos))
```
`pos >= 1 && pos <= top - 1` -> `pos > 0 && pos < top`?

alabuzhev replied:
What I meant is "the expression can be shortened", but it's minor, feel free to ignore.

MKadaner replied:
I see. I was concentrating on the semantics. And speaking about semantics, since in Lua things are counted from 1, having these ones explicit in the code better conveys the meaning of the operation. Also, this pattern is used in the two `case`s above which makes me think that this is the established style here.

All in all, I strongly suggest keeping this expression as is.

MKadaner commented:
I added a comment explaining enigmatic `-3` and addressed one `SonarQube` warning.

All other `SonarQube` warnings are either require rather substantial refactoring (like `using` instead of `typedef` in the entire `plugin.hpp`) or need to be considered in the wider context of HTML documentation formatting. In either case they do not belong in this PR.

MKadaner commented:
Minor updates according to offline discussion with @zg0, rebased to master, and adjusted changelogs and versions.

Ready for review.

HamRusTal commented:
> offline discussion with @zg0

At the annual Worldwide FAR Manager Developers Conference?  :)

alabuzhev commented on far/editor.cpp:
```diff
@@ -3437,6 +3437,86 @@ namespace
+			const auto KeyStr{ AsString(Key) };
+			if (!KeyStr) return std::nullopt;
+
+			if (KeyStr == L"Line")
```
Minor: pls use L"Line"sv here and below.

MKadaner replied:
Will do. Please let me know when you have finished with the comments.

MKadaner replied:
Done!

alabuzhev commented:
Thank you

MKadaner commented:
Thank you!
